src/globus_cli/exception_handling/hooks.py

```diff
@@ -24,32 +24,47 @@ def _pretty_json(data: dict, compact=False) -> str:
     error_class=CLIAuthRequirementsError,
     exit_status=4,
 )
-def handle_internal_auth_requirements(exception: CLIAuthRequirementsError) -> None:
+def handle_internal_auth_requirements(
+    exception: CLIAuthRequirementsError,
+) -> int | None:
     gare = exception.gare
     if not gare:
         click.secho(
             "Fatal Error: Unsupported internal auth requirements error!",
             bold=True,
             fg="red",
         )
-        click.get_current_context().exit(255)
+        return 255
+
+    exit_code: int | None = None
 
     required_scopes = gare.authorization_parameters.required_scopes
     if required_scopes:
-        ret = _concrete_consent_required_hook(exception.message, required_scopes)
-        if ret is not None:
-            click.get_current_context().exit(ret)
+        exit_code = _concrete_consent_required_hook(exception.message, required_scopes)
 
     session_policies = gare.authorization_parameters.session_required_policies
     session_identities = gare.authorization_parameters.session_required_identities
     session_domains = gare.authorization_parameters.session_required_single_domain
     if session_policies or session_identities or session_domains:
-        _concrete_session_hook(
+        ret = _concrete_session_hook(
             exception.message,
             session_policies,
             session_identities,
             session_domains,
         )
+        if exit_code is None:
+            exit_code = ret
+        elif ret is not None:
+            # use bitwise-or if we have two codes to handle
+            # it's guaranteed that if either is 255 the result will be 255, and it's
+            # one of the most information-preserving possible handling paths
+            exit_code |= ret
```

Comment:
No. I do not want to do that.
The definition here is that there are internal hooks which are allowed to indicate an exit status or None.

The fact that this hook happens not to leverage this right now doesn't change the design.

I could have defined a `Protocol` or wrapped these in some object hierarchy to artificially make the protocol explicit, but that's just so much noise relative to the interface being "returns exit code or None".

Reply:
I don't see a way to override this unless the module-level function is swapped out, but this is the CLI, so that's not going to happen. `_concrete_session_hook()` exclusively returns `None` so I'm saying this code can be reverted because `ret` is `None` and this will never run.

Maybe I'm missing something?

Reply:
We talked this over to sync; and I'll share the result here:

I was concerned with making sure we had a paradigm in place for handling a series of operations to handle GAREs, with some kind of well-defined cascading failure semantics. Each step executes serially, but we only aggregate-up the results at the end. Everything we mismatched about was in service of having this kind of setup -- which skips over an important but valuable question of whether cascading through to a later failure or failing early is better, etc.

I'm going to adjust the method for handling invalid/internally-inconsistent GARE data (e.g. ConsentRequired with no scopes listed) to evaluate at the start of the error hook whether or not the data appears to be well-formed (to a higher standard than whether or not it can be parsed -- namely, does it match exactly what the CLI expects). After doing so, we'll either error with a clear fail-whale message or proceed with normal (non-erroring) handling.